pep-0646.rst

```diff
@@ -743,11 +753,220 @@ is available in `cpython/23527`_. A preliminary version of the version
+''''''''''''''''''''''''''
+
+A second approach (the one that most of the examples in this PEP are based around)
+would be to have names constrain axis *type* but not axis *size*.
+
+This would enable us to solve the problem of enforcing shape properties across calls.
+For example:
+
+::
+
+    # lib.py
+    
+    class Batch: pass
+    class Time: pass
+    
+    def make_array() -> Array[Batch, Time]: ...
+    
+    # user.py
+    
+    from lib import Batch, Time
+    
+    # `Batch` and `Time` have the same identity as in `lib`,
+    # so must take array as produced by `lib.make_array`
+    def use_array(x: Array[Batch, Time]): ...
+    
+In many cases, this is the more important thing to verify; we care more about
+which axis is which than what the specific size of each axis is.
+
+It also does not preclude use cases where we wish to describe shape transformations
+where we don't know semantic identity ahead of time. For example, we can write:
+
+::
+
+    K = TypeVar('K')
+    N = TypeVar('N')
+
+    def matrix_vector_multiply(x: Array[K, N], Array[N]) -> Array[K]: ...
+    
+We can then use this with:
+    
+    class Batch: pass
+    class Values: pass
+    
+    batch_of_values: Array[Batch, Values]
+    value_weights: Array[Values]
+    matrix_vector_multiply(batch_of_values, value_weights)
+    # Result is Array[Batch]
+
+How Flexible is this PEP?
+'''''''''''''''''''''''''
+
+The approach described in the previous section is straightforwardly
+compatible with this PEP. It is, essentially, standard parametric subtyping
+as applied to arrays.
+
+What might not be so obvious is that, using ``Literal``, we can *also* apply Python's
+rules for parametric subtyping to the other case, too! For example:
+
+::
+
+    K = TypeVar('K')
+    N = TypeVar('N')
+
+    def matrix_vector_multiply(x: Array[K, N], Array[N]) -> Array[K]: ...
+
+    a: Array[Literal[64], Literal[32]]
+    b: Array[Literal[32]]
+    matrix_vector_multiply(a, b)
+    # Result is Array[Literal[64]]
+
+However, note that the two approaches are mutually exclusive in user code. Users
+can verify size or semantic type but not both.
+
+As of this PEP, we are agnostic about which approach will provide most benefit.
+Since the features introduced in this PEP are compatible with both approaches, however,
+we leave the door open.
+
+Why Not Both?
+'''''''''''''
+
+Consider the following 'normal' code:
+
+::
+
+    def f(x: int): ...
+   
+Note that we have symbols for both the value of the thing (``x``) and the type of
+the thing (``int``). Why can't we do the same with axes? For example, with an imaginary
+syntax, we could write:
+
+::
+
+    def f(array: Array[TimeValue: TimeType]): ...
+    
+This would allow us to access the axis size (say, 32) through the symbol ``TimeValue``
+*and* the type through the symbol ``TypeType``.
+
+This might even be possible using existing syntax, through a second level of parameterisation:
+
+::
+
+   def f(array: array[TimeValue[TimeType]]): ..
```

Comment:
> If tensor has type Tensor[L[32], L[8]], then tensor.shape() has type Tuple[L[32], L[8]]. So, tensor.shape()[0] has the type L[32] and can be used as expected.

Right, but we only have axis _value_ information in there - not axis _type_ information.

I'm imagining something like:

```python3
def zoom_image_by_factor(
    image: Tensor[HeightType[HeightValue], WidthType[WidthValue]],
    zoom_factor: int,
):
  zoomed_image: Tensor[HeightType, WidthType] = lib.resize_image_to_size(
      image,
      (WidthValue * zoom_factor, HeightValue * zoom_factor),  # e.g. (64 * 2, 64 * 2)
  )
  return zoomed_image
```
Here we want access to _both_ i) the actual axis size (`WidthValue`, `HeightValue`), and ii) the axis type, in order to make sure that `resize_image_to_size` is returning a `Tensor` with the axes in the order we expect.

Or maybe just:
```python3
def zoom_image_by_factor(
    image: Tensor[HeightType[HeightValue], WidthType[WidthValue]],
    zoom_factor: int,
) -> Tensor[HeightType[zoom_factor * HeightValue], WidthType[zoom_factor * WidthValue]]:
  ...
```
(Though this involves dependent typing, which is out-of-scope even for Alfonso's proposed extensions, right?)

Reply:
> Right, but we only have axis value information in there - not axis type information.

Yes, the value is something that must be present at runtime. In the above example, it is `image.height` and `image.width`. If there are no such attributes on the tensor, there will be no runtime value for Python to use. Type information is not present at runtime.

> Here we want access to both i) the actual axis size (WidthValue, HeightValue), and ii) the axis type, in order to make sure that resize_image_to_size is returning a Tensor with the axes in the order we expect.

We get (i) the actual axis size from `image.height`, etc. and (ii) the axis type from the tensor type.

In other words, the untyped code for `zoom_image_by_factor` is:

```
    zoomed_image = lib.resize_image_to_size(image, (image.width * zoom_factor, image.height * zoom_factor))
```

As the typechecker, we make sure that, for `image: Tensor[L[32], L[8]]`, `image.width` has the type `L[32]` and `image.height` has the type `L[8]`. That way, we catch any shape errors that might occur. The fully annotated function signature is below:

```
H = TypeVar("H", bound=int)
W = TypeVar("W", bound=int)
N = TypeVar("N", bound=int)

def zoom_image_by_factor(
    image: Tensor[H, W]],
    zoom_factor: N,
) -> Tensor[Mul[H, N], Mul[W, N]]:
    zoomed_image = lib.resize_image_to_size(image, (image.width * zoom_factor, image.height * zoom_factor))
    return zoomed_image
```

Let me know if that makes sense. Overall, I feel like we can remove this section.

Reply:
> As the typechecker, we make sure that, for image: Tensor[L[32], L[8]], image.width has the type L[32] and image.height has the type L[8].

Ahh, maybe that was the bit of context I was missing - I wasn't necessarily thinking `image` would have `.width` and `.height`. And I don't think we should make that assumption in general, since we want to solve this problem in the general case where we're just annotating plain `Tensor`s.

Just to make sure all the context is clear from my end too: in general the reason I'm against annotating only the axis values is that it doesn't guard against cases like this:

```python
H = TypeVar('H', bound=int)
W = TypeVar('W', bound=int)

# This function relies on H and W being the right way around.
# (That is, if H and W are swapped, it'll do vertical line detection instead.)
def horizontal_line_detection(image: Array[H, W]) -> Array[H, W]:
  ...

width = 32
height = 64
image = np.zeros((width, height))  # Oops - axis ordering reversed!
horizontal_line_detection(image)  # Oh no!
```

Ideally we want some way of catching errors like this (implying we should annotate the semantic type of each dimension) while at the same time being able to specify how a function changes the actual shape values (e.g. `N -> Mul[W, N]`).

Reply:
> Ahh, maybe that was the bit of context I was missing - I wasn't necessarily thinking image would have .width and .height. And I don't think we should make that assumption in general, since we want to solve this problem in the general case where we're just annotating plain Tensors.

In my original comment, I'd mentioned `tensor.shape` and `tensor.shape[0]`, which works for arbitrary-sized tensors and does not assume a `height` attribute. In general, we need some runtime value that stores the shape.

> Just to make sure all the context is clear from my end too: in general the reason I'm against annotating only the axis values is that it doesn't guard against cases like this:

Gotcha. That seems to be your main objection to the axis-size notation. 

> if H and W are swapped, it'll do vertical line detection instead

I don't agree. Instead of doing horizontal line detection on a 64x32 image, it will still do horizontal line detection but on a 32x64 image.

Overall, we are trying to catch runtime errors using the typechecker. There is no runtime error in the above code sample. However, there will be an error when *using* the 32x64 image against some 64x32 image and we will indeed catch that:

```
H = TypeVar('H', bound=int)
W = TypeVar('W', bound=int)

def horizontal_line_detection(image: Array[H, W]) -> Array[H, W]:
  ...

width = 32
height = 64
image = np.zeros((width, height))  # Oops - axis ordering reversed!
horizontal_line_detection(image)  # Oh no!

overlay_image = np.zeros((height, width))

# Type error: expected Tensor[L[64], L[32]] but got Tensor[L[32], L[64]].
overlay(image, overlay_image)
```

Let me know if you had an example where axes were switched but were never used against another image in a manner that would cause a runtime error.

Reply:
(Have confirmed with Pradeep offline he's fine with this section now.)